File: in_app_purchases/step_09/firebase-backend/functions/src/purchase-handler.ts
import {ProductData} from "./products";

export abstract class PurchaseHandler {
  async verifyPurchase(
      userId: string,
      productData: ProductData,
      token: string,
  ): Promise<boolean> {
    switch (productData.type) {
      case "SUBSCRIPTION":
        return this.handleSubscription(userId, productData, token);
      case "NON_SUBSCRIPTION":
        return this.handleNonSubscription(userId, productData, token);
      default:
        return false;
    }
  }

  abstract handleNonSubscription(
      userId: string,
      productData: ProductData,
      token: string,
  ): Promise<boolean>;

  abstract handleSubscription(
      userId: string,
      productData: ProductData,
      token: string,
  ): Promise<boolean>;
}

File: in_app_purchases/complete/firebase-backend-dart/lib/purchase_handler.dart
import 'package:firebase_backend_dart/products.dart';

abstract class PurchaseHandler {
  Future<bool> verifyPurchase({
    required String userId,
    required ProductData productData,
    required String token,
  }) async {
    switch (productData.type) {
      case ProductType.subscription:
        return handleSubscription(
          userId: userId,
          productData: productData,
          token: token,
        );
      case ProductType.nonSubscription:
        return handleNonSubscription(
          userId: userId,
          productData: productData,
          token: token,
        );
    }
  }

  Future<bool> handleNonSubscription({
    required String userId,
    required ProductData productData,
    required String token,
  });

  Future<bool> handleSubscription({
    required String userId,
    required ProductData productData,
    required String token,
  });
}
